elastic/eui
# [EuiCode][EuiCodeBlock] Various cleanups and bug fixes

### Summary

closes https://github.com/elastic/eui/issues/5173
closes https://github.com/elastic/eui/issues/5381
closes https://github.com/elastic/eui/issues/5159

This PR primarily separates `EuiCode` and `EuiCodeBlock` from sharing as much unnecessary code/state as it does (since `EuiCode` is dramatically simpler) and DRYs out the code that they do share into utils.

Along the way however I noticed several other bugs and things to clean up - so I strongly recommend [following along by commit](https://github.com/elastic/eui/pull/5379/commits). I should hopefully have left enough context and rationale for certain changes/decisions in my commit messages - please let me know if not!

## QA

- [x] Inline `EuiCode` should look and work the same as before (sans bugfixes)
  - [x] Long code strings now wrap normally 
- [x] `EuiCodeBlock` should look and work the same as before (sans bugfixes)
- Bugfixes
    - Go to https://eui.elastic.co/pr_5379/#/editors-syntax/code
    - Set the theme to Amsterdam
    - In the first `EuiCodeBlock` example, open the playground toggle
    - Set the `overflowHeight` to (e.g.) `150` in order to test virtualization and full screen behavior
    - [x] Set `fontSize` to `l` and confirm the code font changes correctly
    - [x] Set `isVirtualized` to `true` and confirm the line height remains the same on `fontSize="l"`
    - [x] Set `aria-label` & `data-test-subj` to any text and inspect the underlying `<code>` example in the live playground. Confirm the props were correctly passed through
    - [x] Toggle full screen mode, then press the Escape key. Confirm full screen mode is removed
    - [x] Set the `fontSize` and `paddingSize` to `s` and then re-toggle full screen mode. Confirm that the font and padding are `large` in full screen mode
    - Close playground mode
    - [x] Go to the `whiteSpace="pre"` example and click the full screen toggle button. Confirm that the text does not visually overlap the controls and that scrolling all the way to the right reveals the extent of the text.
    - [x] Go to the virtualization example and scroll down to to the middle of the text. Click the full screen button and then close full screen mode. Confirm the non-full-screen code block has not been reset or blanked out.

### Checklist

- [x] Check against **all themes** for compatibility in both light and dark modes

> ⚠️  I don't _think_ this PR contains any breaking changes since `EuiCodeBlockImpl` should have been an internal implementation only, but let me know if you disagree. I also don't think the CSS class changes on `EuiCode` is technically a breaking change, from my understanding of how we've handled CSS changes.
